From: Jian Shen <shenjian15@huawei.com>
Date: Fri, 5 Jan 2018 18:18:11 +0800
Subject: net: hns3: Mask the packet statistics query when NIC is down
Patch-mainline: v4.16-rc1
Git-commit: b875cc379de4cb4bcc7e189998667286baf916b8
References: bsc#1104353 FATE#326415

Update the HNS3_NIC_STATE_DOWN bit when NIC state changes.
When NIC is down, mask the packet statistics for querying
with ifconfig command. It's a common practice.

Signed-off-by: Jian Shen <shenjian15@huawei.com>
Signed-off-by: Peng Li <lipeng321@huawei.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/net/ethernet/hisilicon/hns3/hns3_enet.c |    8 ++++++++
 1 file changed, 8 insertions(+)

--- a/drivers/net/ethernet/hisilicon/hns3/hns3_enet.c
+++ b/drivers/net/ethernet/hisilicon/hns3/hns3_enet.c
@@ -246,6 +246,8 @@ static int hns3_nic_net_up(struct net_de
 	if (ret)
 		goto out_start_err;
 
+	clear_bit(HNS3_NIC_STATE_DOWN, &priv->state);
+
 	return 0;
 
 out_start_err:
@@ -285,6 +287,9 @@ static void hns3_nic_net_down(struct net
 	const struct hnae3_ae_ops *ops;
 	int i;
 
+	if (test_and_set_bit(HNS3_NIC_STATE_DOWN, &priv->state))
+		return;
+
 	/* stop ae_dev */
 	ops = priv->ae_handle->ae_algo->ops;
 	if (ops->stop)
@@ -1134,6 +1139,9 @@ hns3_nic_get_stats64(struct net_device *
 	u64 tx_pkts = 0;
 	u64 rx_pkts = 0;
 
+	if (test_bit(HNS3_NIC_STATE_DOWN, &priv->state))
+		return;
+
 	handle->ae_algo->ops->update_stats(handle, &netdev->stats);
 
 	for (idx = 0; idx < queue_num; idx++) {
